rust/connlib/tunnel/src/io.rs

```diff
@@ -179,6 +218,91 @@ impl Io {
         Ok(())
     }
 
+    pub fn send_dns_query(&mut self, query: dns::RecursiveQuery) {
+        match query.transport {
+            dns::Transport::Udp => {
+                let factory = self.udp_socket_factory.clone();
+                let server = query.server;
+                let bind_addr = match query.server {
+                    SocketAddr::V4(_) => SocketAddr::new(Ipv4Addr::UNSPECIFIED.into(), 0),
+                    SocketAddr::V6(_) => SocketAddr::new(Ipv6Addr::UNSPECIFIED.into(), 0),
+                };
+                let meta = DnsQueryMetaData {
+                    query: query.message.clone(),
+                    server,
+                    transport: dns::Transport::Udp,
+                };
+
+                if self
+                    .dns_queries
+                    .try_push(
+                        async move {
+                            // To avoid fragmentation, IP and thus also UDP packets can only reliably sent with an MTU of <= 1500 on the public Internet.
+                            const BUF_SIZE: usize = 1500;
+
+                            let udp_socket = factory(&bind_addr)?;
+
+                            let response = udp_socket
+                                .handshake::<BUF_SIZE>(server, query.message.as_slice())
+                                .await?;
```

Comment:
To expand on making `UdpSocket::handshake` static, I think here we would need a different factory function probably.

Reply:
Yeah I wanted to avoid having to pass in a 2nd factory function through all layers.